### DISCUSSION TITLE
OAuth2 - am I using the getAccessToken request modifier correctly?

### DISCUSSION THREAD

**Discussion by nikspyratos** [saloonphp/saloon#201] [Q&A]
Heyo 👋 

In my Saloon-based SDK, I need to modify the GetAccessTokenRequest to add a header. In my case the API I'm working with takes the client ID + secret in base64 form in an `Authorization: Basic xxxx` header and not in the request body.

What I'm doing to achieve this is as follows ([link here](https://github.com/nikspyratos/investec-sdk-php/blob/9ebab51962e7799a50cbc3d248cf524078bf2001/src/Connectors/InvestecOAuthConnector.php#L66)):

```
$requestModifier = function (GetAccessTokenRequest $request) {
    $request->withTokenAuth(base64_encode($this->oauthConfig()->getClientId().':'.$this->oauthConfig()->getClientSecret()), 'Basic');
};

$this->getAccessToken($code, $state, $expectedState, $returnResponse, $requestModifier);
```

But when I log the request contents when it gets to `PendingRequest`, `$this->request->headers()` returns an empty `ArrayStore`.

---

My alternative option is to override the trait method completely and replace `GetAccessTokenRequest` with a custom implementation (this does seem to work for getting modified headers), but I just wanted to know if I'm doing something wrong here.

**Comment by nikspyratos** [comment]
Update: Due to other quirks of the API in question I've had to go with overriding `getAccessToken` with a custom `GetAccessTokenRequest` anyway, so I suppose this is more just a question of interest right now.

**Comment by Sammyjo20** [comment]
Hey Nik! 🤠

Perhaps it's because of the way Saloon's authenticators work. At what stage are you dumping the PendingRequest? If you dump it right after, then the middleware responsible for converting the authenticator might not have run yet. When you send the request does it work or does it still fail?
